File: file.cpp
#include <iostream>
#include <fstream>
#include <opencv2/opencv.hpp>
#include <Eigen/Dense>

using namespace std;
using namespace cv;
using namespace Eigen;

int main() {
    // Load the dataset
    MatrixXd dataset;
    ifstream file("Salary_Data.csv");
    string line;
    vector<double> X, y;
    while (getline(file, line)) {
        double x_val, y_val;
        sscanf(line.c_str(), "%lf,%lf", &x_val, &y_val);
        X.push_back(x_val);
        y.push_back(y_val);
    }
    dataset.conservativeResize(X.size(), 2);
    for (int i = 0; i < X.size(); ++i) {
        dataset(i, 0) = X[i];
        dataset(i, 1) = y[i];
    }

    // Split the dataset into the Training set and Test set
    int num_samples = dataset.rows();
    int num_train = num_samples * 2 / 3;
    int num_test = num_samples - num_train;
    MatrixXd X_train = dataset.block(0, 0, num_train, 1);
    MatrixXd y_train = dataset.block(0, 1, num_train, 1);
    MatrixXd X_test = dataset.block(num_train, 0, num_test, 1);
    MatrixXd y_test = dataset.block(num_train, 1, num_test, 1);

    // Define the linear regression model
    MatrixXd X_ones = MatrixXd::Ones(num_train, 1);
    MatrixXd X_train_augmented(num_train, 2);
    X_train_augmented << X_train, X_ones;
    MatrixXd W = (X_train_augmented.transpose() * X_train_augmented).inverse() * X_train_augmented.transpose() * y_train;

    // Predictions
    MatrixXd X_test_ones = MatrixXd::Ones(num_test, 1);
    MatrixXd X_test_augmented(num_test, 2);
    X_test_augmented << X_test, X_test_ones;
    MatrixXd y_pred = X_test_augmented * W;

    // Calculate Mean Squared Error
    double mse = (y_test - y_pred).array().square().mean();

    // Visualize the Training set results
    Mat image(600, 600, CV_8UC3, Scalar(255, 255, 255));
    for (int i = 0; i < num_train; ++i) {
        int x = (X_train(i, 0) - X_train.minCoeff()) / (X_train.maxCoeff() - X_train.minCoeff()) * 600;
        int y = (y_train(i, 0) - y_train.minCoeff()) / (y_train.maxCoeff() - y_train.minCoeff()) * 600;
        circle(image, Point(x, 600 - y), 5, Scalar(0, 0, 255), -1);
    }

    // Visualize the Test set results
    Mat image_test(600, 600, CV_8UC3, Scalar(255, 255, 255));
    for (int i = 0; i < num_test; ++i) {
        int x = (X_test(i, 0) - X_train.minCoeff()) / (X_train.maxCoeff() - X_train.minCoeff()) * 600;
        int y = (y_test(i, 0) - y_train.minCoeff()) / (y_train.maxCoeff() - y_train.minCoeff()) * 600;
        circle(image_test, Point(x, 600 - y), 5, Scalar(0, 0, 255), -1);
    }

    // Show the visualizations
    imshow("Training Set", image);
    imshow("Test Set", image_test);
    waitKey(0);

    cout << "Mean Squared Error: " << mse << endl;

    return 0;
}

File: file.py
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import torch
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

# Import the dataset
dataset = pd.read_csv('Salary_Data.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 1].values

# Split the dataset into the Training set and Test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=1/3, random_state=0)

# Convert data to PyTorch tensors
X_train = torch.tensor(X_train, dtype=torch.float32)
y_train = torch.tensor(y_train, dtype=torch.float32)
X_test = torch.tensor(X_test, dtype=torch.float32)
y_test = torch.tensor(y_test, dtype=torch.float32)

# Define the linear regression model
class LinearRegressionModel(torch.nn.Module):
    def __init__(self):
        super(LinearRegressionModel, self).__init__()
        self.linear = torch.nn.Linear(1, 1)

    def forward(self, x):
        return self.linear(x)

model = LinearRegressionModel()

# Define loss function and optimizer
criterion = torch.nn.MSELoss()
optimizer = torch.optim.SGD(model.parameters(), lr=0.01)

# Train the model
num_epochs = 1000
for epoch in range(num_epochs):
    # Forward pass
    outputs = model(X_train)
    loss = criterion(outputs, y_train)

    # Backward pass and optimization
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

# Predictions
y_pred = model(X_test)

# Calculate Mean Squared Error
mse = mean_squared_error(y_test, y_pred.detach().numpy())

# Visualize the Training set results
plt.figure(figsize=(12, 5))
plt.subplot(1, 2, 1)
plt.scatter(X_train.numpy(), y_train.numpy(), color='red', label='Actual')
plt.plot(X_train.numpy(), model(X_train).detach().numpy(), color='blue', label='Predicted')
plt.title('Salary vs Experience (Training Set)')
plt.xlabel('Years of Experience')
plt.ylabel('Salary')
plt.legend()

# Visualize the Test set results
plt.subplot(1, 2, 2)
plt.scatter(X_test.numpy(), y_test.numpy(), color='red', label='Actual')
plt.plot(X_test.numpy(), y_pred.detach().numpy(), color='blue', label='Predicted')
plt.title('Salary vs Experience (Test Set)')
plt.xlabel('Years of Experience')
plt.ylabel('Salary')
plt.legend()

# Additional data visualization
plt.figure(figsize=(12, 5))
plt.subplot(1, 2, 1)
plt.scatter(X.numpy(), y, color='red')
plt.plot(X_train.numpy(), model(X_train).detach().numpy(), color='blue')
plt.title('Salary vs Experience (Full Data)')
plt.xlabel('Years of Experience')
plt.ylabel('Salary')

# Plot loss vs. epoch
losses = []
for epoch in range(num_epochs):
    outputs = model(X_train)
    loss = criterion(outputs, y_train)
    losses.append(loss.item())

plt.subplot(1, 2, 2)
plt.plot(range(num_epochs), losses)
plt.title('Loss vs. Epoch')
plt.xlabel('Epoch')
plt.ylabel('Loss')

plt.show()

print(f"Mean Squared Error: {mse:.2f}")
